File: List/LRUCache/LRUCache.java
/*
class ListNode {
	int val;
	ListNode next;
	
	ListNode(int val) {
		this.val = val;
		this.next = null;
	}
};
*/
lass Solution {
	public boolean palindromeLinkedList(ListNode head) {
		if(head == null || head.next == null)
			return true;
		
		Deque<Integer> stack = new Deque<Integer>();
		ListNode slow = head;
		ListNode fast = head;
		
		while(fast != null && fast.next != null) {
			stack.push(slow.val);
			slow = slow.next;
			fast = fast.next.next;
		}
		if(fast)
			slow = slow.next;
		
		ListNode curr = slow;
		if(curr != null) {
			if(curr.val != stack.pop())
				return false;
			curr = curr.next;
		}
		return true;
	}	
	
	public boolean palindromeLinkedList2(ListNode head) {
		if(head == null || head.next == null)
			return true;
		
		Deque<Integer> stack = new Deque<Integer>();
		ListNode slow = head;
		ListNode fast = head;
		
		while(fast != null && fast.next != null) {
			stack.push(slow.val);
			slow = slow.next;
			fast = fast.next.next;
		}
		if(fast)
			slow = slow.next;
		
		ListNode rHead = reverse(slow);
		ListNode lCurr = head, rCurr = rHead;
		if(rCurr != null) {
			if(curr.val != lCurr.val)
				reverse(rHead);
			lCurr = lCurr.next;
			rCurr = rCurr.next;
		}
		reverse(rHead);
		return true;
	}
	
	private ListNode reverse(ListNode head) {
		if(head == null)
			return;
		
		ListNode prev = null;
		while(head != null) {
			ListNode after = head.next;
			head.next = prev;
			prev = head;
			head = after;
		}
		return prev;
	}
}

File: List/LRUCache/LRUCache.cc
/*
class ListNode {
	int val;
	ListNode* next;
	
	ListNode(int v):val(v), next(null) {}
};
*/

class Solution {
	public:
		bool LRUCache(ListNode* head) {
			if(head == NULL || head->next == NULL)
				return true;
			
			vector<int> stack = new vector<int>();
 			ListNode* slow = head;
			ListNode* fast = head;
			
			while(fast != NULL && fast->next != NULL) {
				stack.push_back(slow->val);
				slow = slow->next;
				fast = fast->next->next;
			}
			if(fast)
				slow = slow->next;
			
			ListNode* curr = slow;
			if(curr != NULL) {
				if(curr->val != stack.pop())
					return false;
				curr = curr->next;
			}
			return true;
		}

		bool palindromeLinkedList2(ListNode* head) {
			if(head == NULL || head->next == NULL)
				return true;
			
			vector<int> stack = new vector<int>();
 			ListNode* slow = head;
			ListNode* fast = head;
			
			while(fast != NULL && fast->next != NULL) {
				stack.push_back(slow->val);
				slow = slow->next;
				fast = fast->next->next;
			}
			if(fast)
				slow = slow->next;
			
			ListNode* rHead = reverse(slow);
			ListNode* lCurr = head, rCurr = rHead; 
			if(lCurr != NULL) {
				if(lCurr->val != rCurr->val) {
					reverse(rHead);
					return false;
				}
				lCurr = lCurr->next;
				rCurr = rCurr->next;
			}
			reverse(rHead);
			return true;
		}	

	private:
		ListNode* reverse(ListNode* head) {
			if(head == NULL)
				return NULL;
			
			ListNode* prev = NULL;
			while(head) {
				ListNode* after = head->next;
				head->next = prev;
				prev = head;
				head = after;				
			}
			return prev;
		}
};